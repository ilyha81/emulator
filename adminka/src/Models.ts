
import {EDocumentStatus, EOperationStatus} from "./Enums";

/**
 * Интерфейс документа отправляемого при создании операции.
 */
export interface IDocument {
    id: string;
    docType: string;
}

/**
 * Интерфейс для Документа в операции подписания.
 * @prop {EDocumentStatus} status Статус документа.
 * @prop {string} errorCode Код ошибки если она произошла.
 * @prop {string} errorMsg Сообщение об ошибке, если она произошла.
 */
export interface IDocumentServer extends IDocument{
    status: EDocumentStatus;
    errorCode?: string;
    errorMsg?: string;
}

/**
 * Model of Operation on server
 */
export interface IOperationServer {
    id: string;
    status: EOperationStatus,
    timer: number,
    cryptoprofileId: string;
    documents: IDocumentServer[],
    createDate: string,
    dataSourceMode: string
}