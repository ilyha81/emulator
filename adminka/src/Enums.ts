/**
 * Documents LifeCycle Statuses.
 * @prop SENDING Отправлен на подписание.
 * @prop POSTED ХЗ
 * @prop DAY_LIMIT_ERROR ХЗ
 *
 */
export enum EDocumentStatus {
    SENDING = "SENDING",
    POSTED = "POSTED",
    DAY_LIMIT_ERROR = "DAY_LIMIT_ERROR"
}

/**
 * Operation LifeCycle Statuses.
 */
export enum EOperationStatus {
    POSTED_SUCCESS = "POSTED_SUCCESS",
    ANSWER_REFUSED = "ANSWER_REFUSED",
    PROCESSED_SUCCESS = "PROCESSED_SUCCESS",
    CONFIRMED = "CONFIRMED",
    PROCESSED_FAILED = "PROCESSED_FAILED",
    ANSWER_FAILED = "ANSWER_FAILED",
    SENDING = "SENDING",
    POSTED_FAILED = "POSTED_FAILED",
    SENDING_FAILED = "SENDING_FAILED",
}