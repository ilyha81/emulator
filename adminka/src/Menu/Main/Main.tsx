import * as React from "react";
import "./Main.css";

/**
 * Основное рабочее окно.
 */
export const Main: React.FC = (props) => {

    return <div className="Main">
        <div>
            Рабочий экран.
        </div>
        {props.children}
    </div>
};