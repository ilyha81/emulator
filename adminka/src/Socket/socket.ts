import io from "socket.io-client";
import {BehaviorSubject} from "rxjs";

/**
 * Типы сообщений.
 */
export enum ESocketMsg {
    MESSAGE = "MESSAGE",
    SHOW_DATA = "SHOW_DATA",
    TEST = "TEST",
}

export type TSocketMessageCB = (data: string) => void;

/**
 * Функция отрисовки данных в соответствующей секции
 * @param data Данные для отрисовки
 */
const drawData = (data: string) => {
    const section = document.getElementById('show-data');
    console.log(data);
    section && (section.innerText = JSON.stringify(data));
};

/**
 * Сам коннект Сокета.
 */
const socket = io.connect('http://localhost:8000', {transports: ['websocket', 'polling']});

/**
 * Создаем подписчикjd.
 */
const MessageSubscriber = new BehaviorSubject<string>("");
const OperationsSubscriber = new BehaviorSubject<string>("");

/**
 * Создаем рассылку.
 */
socket.on(ESocketMsg.MESSAGE, (data: string) => {
    console.warn("SCOKET: incomming message = " + ESocketMsg.MESSAGE);
    MessageSubscriber.next(data)
});
socket.on(ESocketMsg.SHOW_DATA, (data: string) => {
    console.warn("SCOKET: incomming message = " + ESocketMsg.SHOW_DATA);
    OperationsSubscriber.next(data)
});

export const SocketAPI = {
    /**
     * Создание тестовой операции.
     */
    emitTestOperation: () => {
        socket.emit(ESocketMsg.TEST, {
            documents: [{id: 'asd1', docType: 'RPP'}, {id: 'asd2', docType: 'RPP'}],
            cryptoprofileId: "TEST_PROFILE"
        })
    },

    /**
     * Подписка на Получение сообщения.
     * @param {string} cb Колбэк
     */
    subscribeOnHello: (cb: TSocketMessageCB) => MessageSubscriber.subscribe(cb),

    /**
     * Подписка на Получение операций.
     * @param {string} cb Колбэк.
     */
    subscribeOnOperations: (cb: TSocketMessageCB) => OperationsSubscriber.subscribe(cb),

};

