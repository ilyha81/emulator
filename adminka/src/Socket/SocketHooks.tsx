import { useState, useEffect } from 'react';
import {ESocketConnection, SocketAPI} from './socket';

/**
 * Hook subscribes on Socket connection status.
 */
export const useSocketConnectionStatus = ():ESocketConnection  => {
    const [socketStatus, setSocketStatus] = useState<ESocketConnection>(ESocketConnection.INITIAL);

    useEffect(()=>{
        const connected = SocketAPI.subscribeOnConnectionStatus(setSocketStatus);

        return connected.unsubscribe
    },[]);

    return socketStatus
}