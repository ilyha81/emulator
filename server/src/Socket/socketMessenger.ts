import {Timer} from '../Timer/interval';
import * as socketio  from "socket.io";
import * as core from "express-serve-static-core";
import {IOperationView} from "../Operation/operation";
import {Server as Socket} from 'socket.io';

/**
 * Типы сообщений.
 */
export enum ESocketMsg {
    MESSAGE = "MESSAGE",
    SHOW_DATA = "SHOW_DATA",
    TEST = "TEST",
}

/**
 * Проверка занятых портов.
 */
const ports: Set<number> = new Set;

export class SocketMessenger {
    socket: socketio.Server;

    static isPortUsed = (port: number) => ports.has(port);

    constructor (socket: Socket, getOperations: () => IOperationView[]) {
        this.socket = socket;

        this.socket.on('connection', (client: socketio.Socket) => {
            console.log('Клиент подключился');

            client.emit(ESocketMsg.MESSAGE, "ПОДКЛЮЧЕНИЕ УСПЕШНО !");
        });
    }

    /**
     * Обработка приема сообщений.
     * @param type
     * @param cb
     */
    on = (type: string, cb: (client: any) => any) => {
        this.socket.on(type, cb)
    };

    emitMsg = (type: string, msg: object) => {
        if (typeof msg === "string") {
            this.socket.emit(type, msg);
        } else {
            this.socket.emit(type, JSON.stringify(msg))
        }
    }
}
