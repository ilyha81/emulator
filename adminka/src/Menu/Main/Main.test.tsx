import * as React from 'react';
import {cleanup, render, screen} from '@testing-library/react';
import * as Hooks from "../../Socket/SocketHooks";

import {Main} from './Main';
import {IOperationServer} from "../../Models";
import {EDocumentStatus, EOperationStatus} from "../../Enums";


/**
 * Test data.
 */
const mockTestOperation: IOperationServer = {
    id: 'testId',
    createDate: '2021-03-01, 17:00:01',
    cryptoprofileId: 'testCryptoprofile',
    dataSourceMode: 'testDataSourceMode',
    documents: [{id: 'docId', docType: 'docType', status: EDocumentStatus.POSTED}],
    status: EOperationStatus.POSTED_SUCCESS,
    timer: 100
};

beforeEach(()=> {
    cleanup();
});

describe('Testing Main Component', () => {

    test('Main is rendered successfully', async () => {
        const { container } = render(<Main/>);
        expect(container).toMatchSnapshot();
    });

    test(' Main renders its name', async () => {
        const { rerender } = render(<Main/>);
        expect(screen.queryByRole('heading')).toBeNull();

        const VASYA = 'VASYA';
        rerender(<Main name={VASYA}/>);
        screen.getByRole('heading');
        expect(screen.getByText(VASYA)).toHaveClass('badge','badge-dark');
    });

    test('Main renders children', async () => {
        const VASYA = 'VASYA';
        const PETYA = 'PETYA';
        const Child: React.FunctionComponent<{str: string}> = ({str}) => {
            return <div>{str}</div>
        };

        render(
            <Main>
                <Child key={VASYA} str={VASYA}/>
                <Child key={PETYA} str={PETYA}/>
            </Main>);
        screen.getByText(VASYA);
        screen.getByText(PETYA);
    });

    test('If items passed - render item', async () => {
        const mockFn = jest.fn(()=> [mockTestOperation]);
        //jest.mock("../../Socket/SocketHooks", ()=>({useSocketGetOperations: mockFn}));
        jest.spyOn(Hooks, 'useSocketGetOperations').mockImplementation(mockFn);
        render(<Main/>);
        await expect(screen.findByText(mockTestOperation.cryptoprofileId)).not.toBeNull();
        expect(mockFn).toHaveBeenCalledTimes(1);
        expect(mockFn).toBeCalledWith();
    });
});