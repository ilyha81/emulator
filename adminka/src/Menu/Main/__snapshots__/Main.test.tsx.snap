// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Main Component Main is rendered successfully 1`] = `
<div>
  <div
    class="Main"
  >
    <div>
      Рабочий экран.
    </div>
  </div>
</div>
`;
